### ISSUE TITLE
Feature: create transfers from directories

### ISSUE THREAD

**Issue by jorikvankemenade** [artefactual-labs/enduro#6]
**Current situation**
Currently, the basic filesystem watcher cannot handle directories. If I copy the `BagTransfer` from the Archivematica sample transfers to my watched directory I get the following error:

```shell
{
 Reason:non retryable error,                                                                                                      
 Details:[
    "Error downloading blob: error copying blob: blob (key \"BagTransfer-6\") (code=Unknown): read /eos/workspace/o/oais/archivematica-qa/file-drop/BagTransfer-6: is a directory"
 ], 
 ScheduledEventId:12, StartedEventId:13, 
 Identity:12240@archivematica-enduro-dev.cern.ch@global
}      
```

**Wanted situation**
It would be nice if we could add the option that if the filesystem watcher encounters a directory that it copies the contents of the directory to the objects folder of the new transfer. Or, in the case of the `BagTransfer`, that we can even configure a workflow for bagits. 

Would it be simple to add each of these cases? And if so is it possible to give me some pointers to add either one of them?

**Comment by sevein** [comment]
To provide some context, I chose to have watchers identify [blobs](https://gocloud.dev/howto/blob/) as opposed to files/directories because it's easy to download them from any machine where I have an Enduro worker deployed for actual processing, although we haven't introduced processing workers yet.

I had to add filesystem watchers because our client decided to share their transfers for now via NFS, but they archive their transfers so I didn't run into the problem that you described.

I've outlined a potential plan:

* Watchers emit [BlobEvent](https://github.com/artefactual-labs/enduro/blob/3bbd48a88dc0cdb4f00fe9dbef63a22f4900bf15/internal/watcher/event.go#L18-L33) objects. That struct could have a new `IsDir bool` property.
* From [filesystemWatcher.Watch](https://github.com/artefactual-labs/enduro/blob/3bbd48a88dc0cdb4f00fe9dbef63a22f4900bf15/internal/watcher/filesystem.go#L89-L99), you could set it to `true` when `os.Stat` says so.
* [processingWorkflow](https://github.com/artefactual-labs/enduro/blob/3bbd48a88dc0cdb4f00fe9dbef63a22f4900bf15/internal/workflow/processing.go#L82) receives the event. As part of the workflow you could evaluate `IsDir` to run a different activity, or you could evaluate it from `DownloadActivity` and do it from there.

**Comment by jorikvankemenade** [comment]
> I had to add filesystem watchers because our client decided to share their transfers for now via NFS, but they archive their transfers so I didn't run into the problem that you described.

So just for clarification, the client only works with a use case where a single file is a single transfer? So there are no complex or structured transfers? 

> I've outlined a potential plan:

Thanks for the effort, I'll have a closer look tomorrow.

**Comment by sevein** [comment]
> So just for clarification, the client only works with a use case where a single file is a single transfer? So there are no complex or structured transfers?

That's the case. The structure of the transfer is actually pretty sophisticated but it's packaged as a tarfile.

**Comment by jorikvankemenade** [comment]
@sevein thank you for the additional information and explanation. I think that at this point it is easier for us to comply with the existing ecosystem by just uploading tars. 

The main reason for this is that I think that the blob abstraction, albeit very nice, is not completely suitable for directories on file systems. The case I am worried about is that of a directory containing many files. When copying this directory to the watched location, the filesystem starts with creating the folder and copying the contents. As soon as the target location is created the filesystem watcher will detect this as an event. This triggers a new, and incomplete, transfer in Archivematica.

In the blob abstraction, this is tackled by modeling the writing of an entire blob as a single transaction. Only after the copying is done, the blob is committed. Unfortunately, this will not work for POSIX move or copy.

I would still very much like this feature to be added to Enduro, but at the moment I don't see a clean solution for this. My suggestion is to keep this ticket open for some additional discussion at some later stage

**Comment by sevein** [comment]
IIRC, Storage Service tries to work around that problem with a staging directory that is hosted by the same filesystem. It feels like a hack though. Honestly I don't know what's a good solution neither, but I really like how a blobstore hides a lot of the complexities of filesystems in distributed environments.

MinIO is great, easy to operate and very scalable. Would it be possible for you to try the MinIO approach? It has some good advantages. The events are also persisted, so you can't miss them if Enduro is down. Also, I'm planning to add Enduro workers and at that point you may awnt to deploy them across multiple machines, e.g. imagine a deployment with k8s and 10 replicas, not having to worry about shared filesystems, because workers will receive signed URLs from workflows in order to download the blobs directly from blobstore.

**Comment by jorikvankemenade** [comment]
>  Honestly I don't know what's a good solution neither, but I really like how a blobstore hides a lot of the complexities of filesystems in distributed environments.

I completely agree! The problem is not with the blobstore model, it's a result of how copying files and directories work in a regular filesystem.

> Would it be possible for you to try the MinIO approach?

I guess this would be possible. But at CERN we already have multiple storage systems deployed. Including blob storage (CephFS and S3). So I think it would make more sense for us to integrate Enduro and one of those storage systems instead of deploying another option.

However, I think in the end this all creates additional usability for the users of the service, the basic functionality stays the same. So for now, the "simplest" think we can do is just create tars for our transfers. I think this is also the "safest" option, with Enduro development being on full speed.

> e.g. imagine a deployment with k8s and 10 replicas

You mean to replicate the enduro workers, so that even if some of them go down you still have your "watchers" running and can still process events and send work to your Archivematica pipelines?

**Comment by sevein** [comment]
Marked as `wontfix`, but leaving the issue open until we can document this limitatoin in filesystem watchers.

**Comment by scollazo** [comment]
This has been addressed at https://github.com/artefactual-labs/enduro/pull/357
